### PR TITLE
fix: improve form is-visible-field check performace

### PR DESCRIPTION
**Pull request by connoratrug**
Fix large form performance, the current form feels unresponsive when interaction with the form ( typing , clicking ) with the form. Using the browser performance tab it could be seen that on each form interaction ( mouse-click, key-press) all of the visibility expressions got executed, making the form feel unresponsive / sluggish  

### What are the main changes you did
Replace the loopt that re-evaluated all columns is-visible check with a signals based approach.

- On form init, setup computed functions ( signals ) for each field that has a visability function.
- Bind the params used in the visability function to the computed values, this sets up observables for the values used in the visability function.
  - if a value that is used in a visibility function is updated, it now calls its observers and triggers re-evaluation of the visibility function ( but only for those fields that are dependant on the change).

- Remove explicit revalidate calls 
- Update Legend component and model to work with reactive model
- Split FormEditModal into Modal and Form, to remove the need for manual reset of form data 

- [x] fix bindings to the form 
- [x] remove all computed side effects
- [x] replace if-else n params block with more dynamic solution 
- [x] add performance test to watch for regressions 

### How to test
- explain here what to do to test this (or point to unit tests)

### Checklist
- [x] updated docs in case of new feature
- [x] added/updated tests
- [x] added/updated testplan to include a test for this fix, including ref to bug using # notation

### CODE REVIEW

File: apps/metadata-utils/src/types.ts [hunk 1]
```diff
@@ -110,13 +111,20 @@ export interface IFieldError {
...
-export interface IFormLegendSection {
-  label: string;
+interface LegendItem {
   id: string;
-  isActive?: boolean;
+  label: string;
   type: HeadingType;
-  section?: string;
-  errorCount: number;
+  errorCount: ComputedRef<number>;
```
Comment (mswertz) [inline apps/metadata-utils/src/types.ts]: inpiring approach

File: data/_models/specific/petstore.csv [hunk 2]
```diff
@@ -27,4 +27,6 @@ User,,,email,email,,,,,,,,,,,,foaf:mbox,Petstore,,,
 User,,,password,,,,,,,,,,,,,,Petstore,,,
 User,,,phone,,,,,,,,,,,,,foaf:phone,Petstore,,,
 User,,,userStatus,int,,,,,,,,,,,,,Petstore,,,
-User,,,pets,multiselect,,,,Pet,,,,,,,,http://example.com/petstore#hasPets,Petstore,,,
\ No newline at end of file
+User,,,pets,multiselect,,,,Pet,,,,,,,,http://example.com/petstore#hasPets,Petstore,,,
+User,,,address,heading,,,,,,,,,,,,,Petstore,,,
+User,,,street,,,,,,,,,,,lastName,,,Petstore,,,
```
[the remark this answers is not in the document]
Comment (mswertz) [reply data/_models/specific/petstore.csv]: not yet hiding when street is hiding :-)

**Review by mswertz — COMMENTED** [review]
Impressive. I did first code review and functional review. Overall I noticed we lack simple data model to test some edge cases, see below.

First functional observations, I tested against
- pet store
- catalogue.Resources
- registry.subject

findings:
- [x] should be noticibly faster on visible than before (no warnings anymore) 
- [ ] menu should highlight section attached to topmost visible column (regression)
        _agree that we can do better, but suggest different dom setup, and implement it on other pr_ 
- [ ] should hide headings if no visible fields (unclear if it works now)
       _agree that we can do better, but suggest different dom setup, and implement it on other pr_ 
- [ ] should hide section if no visible headings or fields  (unclear if it works now)
     _agree that we can do better, but suggest different dom setup, and implement it on other pr_ 
- [ ] errors should only be counted on visible columns (invisible columns should be emptied before sending to the server)
 (unclear) _this is part of the pr, can you add an example where this is not working_

off topic:
- [ ] save validation takes still soo long in 'subject' there should be spinner
- [ ] old version gives different number of error fields, don't know what correct
- [ ] error message says '16 fields require attention before you can save this cohort' should be '... this record' instead of cohort


Example of unwanted behavior

Image

Example of unclear behavior, in previous version we only saw the below in current PR much  more. Probably all to do with headings.

Image

Example address heading should be hidden because no fields

Image

This should only state 'quantity should be >= 1'

Image

**Review by mswertz — APPROVED** [review]
